### PR TITLE
Function deref fix

### LINKED ISSUE #530: ESBMC is unable to dereference function pointer cast for void parameters
ESBMC is not able to detect the bug in the following program:

```c
#include <assert.h>

void *foo() {
  assert(0);
  return 0;
}

void call_function(
    void *(*start_routine)(void*),
    void *arg) {
  (*start_routine)(arg);  
}


int main() {
  call_function(&foo, 0);
  return 0;
}
```

By replacing the signature of `foo` to `void *foo(void*)` ESBMC will find the bug. Executing this program with clang causes the program to fail.

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
Should close #530

### CODE REVIEW

**Review by mikhailramalho — APPROVED** [review]
Ouch!

I wonder the impact of this patch in categories like the linux drivers.

**Review by lucasccordeiro — APPROVED** [review]
LGTM.

File: src/goto-symex/symex_function.cpp [hunk 1]
```diff
@@ -405,7 +405,9 @@ void goto_symext::symex_function_call_deref(const expr2tc &expr)
       return true;
 
     const code_type2t &ct = to_code_type(sym->type);
-    if(ct.arguments.size() != call.operands.size())
+
+    // Passing a list of arguments for a (void) function is legal
```
Comment (fbrausse) [inline src/goto-symex/symex_function.cpp:409]: Nitpick: for `()`, not for `(void)`. Also this only applies to C, not to C++.

**Comment by lucasccordeiro** [comment]
This PR overlaps with https://github.com/esbmc/esbmc/pull/582.

**Pull request closed without merging**